USER-BOCS

mdelyser opened the pull request:
## Purpose

This is part of the BOCS (Bottom-up Open-source Coarse-graining Softare) package (link provided below).

The USER-BOCS package implements a modified barostat for simulating Coarse-Grained models that accurately reproduce structural properties (RDFs) as well as thermodynamic properties (PV Equations of state). The BOCS package includes an old copy of LAMMPS with the same functionality, although done in a way incompatible with the requirements for a user package. It has been rewritten for USER-BOCS to comply with all of the requirements given in section 10.15 "Submitting new feature for inclusion in LAMMPS" on the lammps website.

For a detailed explanation of the barostat modifications, see our published works (links at the end).

## Author(s)

USER-BOCS was written by Nicholas J H Dunn and Michael R DeLyser, both from The Pennsylvania State University

## Backward Compatibility

Nothing in USER-BOCS should break backward compatibility.

## Implementation Notes

The modified barostat was originally included as a modification to fix_nh.cpp to check for a new pressure coupling type of "cgiso". To be included as a user package, we couldn't modify the core LAMMPS file. Accordingly, we copied fix_nh.cpp, renamed it, and removed the extra content we didn't need. The modified barostat is now used via the "fix bocs" command instead of the "fix npt" command. 


For USER-BOCS to work, two lines need to be added to compute.h. We included a file "MOD_COMPUTE_H" that is a copy of compute.h that includes these lines. The two lines are also given in the README file. We assumed that, if this package is accepted, you would be willing to add these lines in the core copy of compute.h - we see that other user packages have additions to the core files, e.g. USER-DPD has short sections in both atom.cpp and atom.h.


Correctness was verified for the USER-BOCS package by comparing results of simulations to those published using the old modified LAMMPS distribution included in the BOCS package.

## Post Submission Checklist

_Please check the fields below as they are completed_
- [X] The feature or features in this pull request is complete
- [X] Suitable new documentation files and/or updates to the existing docs are included
- [X] One or more example input decks are included
- [X] The source code follows the LAMMPS formatting guidelines

## Further Information, Files, and Links

The BOCS package is located at: https://github.com/noid-group/BOCS
Published works:
https://aip.scitation.org/doi/10.1063/1.4937383
https://aip.scitation.org/doi/10.1063/1.4952422
https://pubs.acs.org/doi/abs/10.1021/acs.jpcb.7b09993

rbberger commented on doc/src/USER/atc/man_fix_bocs.html:
```diff
@@ -0,0 +1,301 @@
+
```
Why is this file in atc? Btw, please do not commit generated HTML files.

rbberger commented on src/USER-BOCS/fix_bocs.cpp:
```diff
@@ -0,0 +1,2703 @@
+             "basis types: analytic linear_spline cubic_spline",arg[iarg]);
+         error->all(FLERR,errmsg);
+      } // END NJD MRD  
+    } /*else if (strcmp(arg[iarg],"aniso") == 0) {
```
Why all this uncommented code?

rbberger commented on src/USER-BOCS/MOD_COMPUTE_H:
```diff
@@ -0,0 +1,207 @@
+  virtual void set_arrays(int) {}
+
+// NJD MRD Adding these lines is the only modification needed for USER-BOCS
+  virtual void send_cg_info(int, int, double*, int, double) {}
```
This seems unnecessary. Instead of changing the base class `Compute`, why don't you just store `pressure` as `ComputePressureBocs*` in class `FixBocs`? One doesn't work without the other anyway. So you should try to find a compute of type `ComputePressureBocs` and verify if it has a compatible style and do a type cast from `Compute*` to `ComputePressureBocs*`.

akohlmey commented:
Thanks for your contribution.

Can you explain a bit more in detail the relation between the BOCS package and this pull request?
It all seems a bit confusing. From the outset it looks like BOCS depends primarily on Gromacs, but then also has a LAMMPS component. So is adding the package to LAMMPS giving people any functionality, that is of use without the rest (and gromacs)?

Since your BOCS repository already bundles a customized LAMMPS version, what is the rationale to having your code now included in LAMMPS, rather than simply updating the LAMMPS version in your distribution? while it is generally not encouraged to have changes to the core, it is quite possible to have a few lines added to `FixNH` if that helps to avoid massive code replication. i presume, this is for parsing the arguments and passing them to the various actual fixes derived from FixNH? @rbberger already mentioned, that the changes to compute.h are not necessary. There are plenty of examples in the LAMMPS code base for handling scenarios like this.

mdelyser commented:
Thank you for your comments, suggestions, questions, and consideration. Below, I try to specifically address everything.

@rbberger 

My mistake on the .html files. I deleted them.

I got rid of the changes to compute.h and now cast pressure when calling those functions. Now I internally call the compute for the modified barostat and have the fix_bocs command use it, so the three extra commands needed before are down to one - to report the modified pressure, the thermo_modify command is still required.

We have only tested the modified barostat with isotropic pressure coupling. Originally, I commented out large sections of code instead of deleting them because they deal with non-isotropic pressure coupling, and I wanted to leave in the option of adding support for this in the future. However, if someone wants to implement this in the future, it might be best for them to base their implementation off of the most recent version of fix_npt.cpp instead of the older version used. Additionally, they would have to modify the pressure tensor instead of just the scalar pressure like is currently done. So, I deleted the commented code.

@akohlmey 

BOCS uses GROMACS trajectory files for force matching and the force matching portion of the code must be compiled with GROMACS to make use of their data structures and functions (although we’re working on a newer version of BOCS where we remove the dependence on GROMACS and add support for more file types). Even without the upcoming version of the force matching code, however, I would say that adding the package to LAMMPS still gives people functionality without needing the rest of the BOCS package. There are a variety of methods/software available for parameterizing bottom up coarse-grained models, e.g. the VOTCA package implements Boltzmann Inversion, Iterative Boltzmann Inversion, and Force matching, and the Voth group from Chicago has a github release of their force matching implementation. However, to the best extent of my knowledge, software to use the modified barostat originally described by Das and Andersen ( https://aip.scitation.org/doi/full/10.1063/1.3394862 ) and used by Dunn and Noid for their iterative pressure matching method is not available anywhere, which is why we originally modified LAMMPS to implement it. Although the papers say step 1 is to use the MSCG method to parameterize the coarse-grained force field and step 2 is to follow it up with iterative pressure matching, there’s no reason that any of the other bottom up methods couldn’t be used as step 1 to parameterize the CG interactions. Anyone could, in practice, take any CG force field that reproduces CG RDFs in an NVT simulation and then iteratively pressure match it so it could then be simulated in the NPT ensemble.  

We could, in theory, update the LAMMPS version in our distribution. However, it is likely that anyone interested in using our software already has LAMMPS and possibly not the same version we have in our distribution. If the USER-BOCS package is among the other user contributed packages in the version of LAMMPS they already have, it will be much easier than having to install a second version of LAMMPS for use only with the USER-BOCS functionality. 

Thank you for your time,
Michael DeLyser

rbberger commented:
@mdelyser Thanks for working through the issues.

akohlmey commented:
FYI, I had to merge in recent changes from upstream and adapt the code to it. Also, there were several issues that needed taking care of:
- internal style names (the undocumented pressure compute) are all upper case
- error messages and comments would often refer to `fix nvt/npt/nph`. so this is all `fix bocs`
- removal of redundant example folders
- integration in the manual (html and pdf)
- integration into the (standard) build system
- plug some minor memory leaks
- use force->numeric() or force->inumeric() for parsing parameter numbers
- adjust example filenames and what is output by default to match conventions used elsewhere and record reference log files

This seems to be ready to merge, which should happen in a few days.
